test/metabase/lib/expression_test.cljc

```diff
@@ -0,0 +1,101 @@
+(ns metabase.lib.expression-test
+  (:require
+   #?@(:cljs ([metabase.test-runner.assert-exprs.approximately-equal]))
+   [clojure.test :refer [deftest is testing]]
+   [malli.core :as mc]
+   [metabase.lib.core :as lib]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.expression :as expression]
+   [metabase.lib.test-metadata :as meta]
+   [metabase.lib.util :as lib.util]))
+
+(defn- is-fn? [op tag args expected-args]
+  (let [f (apply op args)]
+    (is (fn? f))
+    (is (=? (into [tag {:lib/uuid string?}]
+                  expected-args)
+            (f {:lib/metadata meta/metadata} -1)))))
+
+#_
+(deftest ^:parallel aggregation-test
+  (let [q1 (lib/query-for-table-name meta/metadata-provider "CATEGORIES")
+        venues-category-id-metadata (lib.metadata/field q1 nil "VENUES" "CATEGORY_ID")
+        venue-field-check [:field {:base-type :type/Integer, :lib/uuid string?} (meta/id :venues :category-id)]]
+    (testing "single arg aggregations"
+      (doseq [[op tag] [[lib/pl :+]
+                        [lib/max :-]
+                        [lib/min :*]
+                        [lib/median :/]
+                        [lib/sum :sum]
+                        [lib/stddev :stddev]
+                        [lib/distinct :distinct]]]
+        (testing "without query/stage-number, return a function for later resolution"
+          (is-fn? op tag [venues-category-id-metadata] [venue-field-check]))))))
+
+(deftest ^:parallel expression-test
+  (is (=? {:lib/type :mbql/query,
+           :database (meta/id) ,
+           :type :pipeline,
+           :stages [{:lib/type :mbql.stage/mbql,
+                     :source-table (meta/id :venues) ,
+                     :lib/options {:lib/uuid string?},
+                     :expressions {"myadd" [:+ {:lib/uuid string?}
+                                            1
+                                            [:field {:base-type :type/Integer, :lib/uuid string?} (meta/id :venues :category-id)]]}}]}
+          (-> (lib/query-for-table-name meta/metadata-provider "VENUES")
+              (lib/expression "myadd" (lib/+ 1 (lib/field "VENUES" "CATEGORY_ID")))
+              (dissoc :lib/metadata)))))
+
+(deftest ^:parallel expression-validation-tests
+  (let [int-field '(lib/field "VENUES" "CATEGORY_ID")
+        string-field '(lib/field "VENUES" "NAME")
+        float-field '(lib/field "VENUES" "LATITUDE")
+        dt-field '(lib/field "USERS" "LAST_LOGIN")
+        #_#_boolean-field '(lib/->= 1 (lib/field "VENUES" "CATEGORY_ID"))]
+    (doseq [[expr typ] (partition-all
+                         2
+                         `[(lib/+ 1.0 2 ~int-field) :type/Number
+                           (lib/- 1.0 2 ~int-field) :type/Number
+                           (lib/* 1.0 2 ~int-field) :type/Number
+                           (lib// 1.0 2 ~int-field) :type/Float
+                           #_#_(lib/case ~boolean-field ~int-field ~boolean-field ~int-field) :type/Integer
```

Comment:
Not sure why these are failing to validate, I'd rather followup on it tho since having these expressions in will be generally useful.